Aflores plymouth (#234)
osisoft/Edge-Data-Store-Docs: content/shared-content/shared-content/configuration/buffering.md

---
uid: BufferingConfiguration
---

# Buffering

You can configure PI adapters to buffer data egressed from the adapter to endpoints. Configure buffering through the buffering configuration parameters in the system configuration.

> **Note:** OSIsoft recommends that you do not modify the default buffering location unless it is necessary. Changes to the buffering configuration parameters only take effect during adapter service startup.

Reviewer: ```suggestion
**Note**: OSIsoft recommends that you do not modify the default buffering location unless it is necessary. Changes to the buffering configuration parameters only take effect during adapter service startup.
```

## Configure buffering

Complete the following steps to configure buffering. Use the `PUT` method in conjunction with the `http://localhost:5590/api/v1/configuration/system/buffering` REST endpoint to initialize the configuration.

> 1. Using a text editor, create an empty text file.
> 
> 2. Copy and paste an example configuration for buffering into the file.
> 
>     For sample JSON, see [Examples - Retrieve the buffering configuration](#examples).
> 
> 3. Update the example JSON parameters for your environment.
> 
>     For a table of all available parameters, see [Buffering parameters](#buffering-parameters).
> 
> 4. Save the file. For example, as `ConfigureBuffering.json`.
> 
> 5. Open a command line session. Change directory to the location of `ConfigureBuffering.json`.
> 
> 6. Enter the following cURL command (which uses the `PUT` method) to initialize the buffering configuration.

Reviewer: ```suggestion
1. Using a text editor, create an empty text file.

1. Copy and paste an example configuration for buffering into the file.

    For sample JSON, see [Examples - Retrieve the buffering configuration](#examples).

1. Update the example JSON parameters for your environment.

    For a table of all available parameters, see [Buffering parameters](#buffering-parameters).

1. Save the file. For example, as `ConfigureBuffering.json`.

1. Open a command line session. Change directory to the location of `ConfigureBuffering.json`.

1. Enter the following `curl` command (which uses the `PUT` method) to initialize the buffering configuration.
```

    ```bash
    curl -d "@ConfigureBuffering.json" -H "Content-Type: application/json" -X PUT "http://localhost:5590/api/v1/configuration/system/buffering"
    ```

> **Note:** If you installed the adapter to listen on a non-default port, update `5590` to the port number in use. For a list of other REST operations you can perform, like updating or replacing a buffering configuration, see [REST URLs](#rest-urls).

Reviewer: ```suggestion
**Note**: If you installed the adapter to listen on a non-default port, update `5590` to the port number you used. For a list of other REST operations you can perform, like updating or replacing a buffering configuration, see [REST URLs](#rest-urls).
```

| **BufferLocation**   | Required  | `string` | Defines the location of the buffer files. Absolute paths are required. Consider the access-control list (ACL) when you set this parameter. **BufferLocation** is used to buffer files when **EnablePersistentBuffering** is `true`. <br><br> Allowed value: Valid path to a folder location in the file system <br> Default value: <br> **Windows:** _%ProgramData%\OSIsoft\Adapters\\{AdapterInstance}\Buffers_ <br> **Linux:** _/usr/share/OSIsoft/Adapters/{AdapterInstance}/Buffers_ |

> <sup>1</sup> **Buffering to disk** - disk is only used if required; <br>

Reviewer: ```suggestion
<sup>1</sup> **Buffering to disk**: Disk is only used if required; <br>
```

> - Data is only written to the disk buffer if it is queued in the memory buffer for more than 5 seconds.
> - The **MaxBufferSizeMB** is applied per configured endpoint, except the health endpoint.<br>
> - An adapter creates 20 MB buffer files that are stored in **BufferLocation**.<br>
> - When **MaxBufferSizeMB** is reached, the oldest buffer file is deleted and a new buffer file is created.<br>
> - The health endpoint is fixed at 20 MB. When the health endpoint buffer file becomes full, a new buffer file is created and the previous buffer file is deleted.
> 
> **Note:** The following rules apply in case of an error when creating a new buffer file:
> 
> - Attempt to delete oldest buffer file and retry.
> - If unable to buffer, errors are logged to indicate data loss.
> - If a buffer file is corrupted, an attempt is made to recover individual records and any failure to recover records is logged.

Reviewer: ```suggestion
- Data is only written to the disk buffer if it is queued in the memory buffer for more than 5 seconds.

- The `MaxBufferSizeMB` is applied per configured endpoint, except the health endpoint.

- An adapter creates 20 MB buffer files that are stored in `BufferLocation`.

- When `MaxBufferSizeMB` reaches its limit, the oldest buffer file is deleted and a new buffer file is created.

- The health endpoint is fixed at 20 MB. When the health endpoint buffer file becomes full, a new buffer file is created and the previous buffer file is deleted.

**Note**: The following rules apply in case of an error when creating a new buffer file:

- Attempt to delete oldest buffer file and retry.

- If unable to buffer, errors are logged to indicate data loss.

- If a buffer file is corrupted, an attempt is made to recover individual records and any failure to recover records is logged.
```

<sup>2</sup> **Buffering only to memory**:<br>

> - The **MaxBufferSizeMB** is applied per configured endpoint except the health endpoint.<br>
> - When **MaxBufferSizeMB** is reached, the oldest messages in the memory buffer are removed. Depending on the size of a new message, several old messages may be removed.<br>
> - The health endpoint is fixed at 20 MB. When the health endpoint buffer file becomes full, the oldest messages in the memory buffer are removed and new messages are added.

Reviewer: ```suggestion
- The **MaxBufferSizeMB** is applied per configured endpoint except the health endpoint.

- When **MaxBufferSizeMB** is reached, the oldest messages in the memory buffer are removed. Depending on the size of a new message, several old messages may be removed.

- The health endpoint is fixed at 20 MB. When the health endpoint buffer file becomes full, the oldest messages in the memory buffer are removed and new messages are added.
```

> - The **MaxBufferSizeMB** is applied per configured endpoint except the health endpoint.<br>
> - When **MaxBufferSizeMB** is reached, the oldest messages in the memory buffer are removed. Depending on the size of a new message, several old messages may be removed.<br>
> - The health endpoint is fixed at 20 MB. When the health endpoint buffer file becomes full, the oldest messages in the memory buffer are removed and new messages are added.

Reviewer: ```suggestion
- The `MaxBufferSizeMB` is applied per configured endpoint except the health endpoint.

- When `MaxBufferSizeMB` is reached, the oldest messages in the memory buffer are removed. Depending on the size of a new message, several old messages may be removed.

- The health endpoint is fixed at 20 MB. When the health endpoint buffer file becomes full, the oldest messages in the memory buffer are removed and new messages are added.
```
